pkg/storage/bloom/v1/tokenizer.go

```diff
@@ -1,15 +1,43 @@
 package v1
 
 import (
+	"fmt"
 	"unicode/utf8"
 
+	"github.com/grafana/loki/pkg/push"
 	iter "github.com/grafana/loki/v3/pkg/iter/v2"
 )
 
 const (
 	MaxRuneLen = 4
 )
 
+type StructuredMetadataTokenizer struct {
+	prefix string
+	buf    []string
+}
+
+func NewStructuredMetadataTokenizer(prefix string) *StructuredMetadataTokenizer {
+	return &StructuredMetadataTokenizer{
+		prefix: prefix,
+		buf:    make([]string, 6),
+	}
+}
+
+// Tokens implements the NGramBuilder interface
+func (t *StructuredMetadataTokenizer) Tokens(kv push.LabelAdapter) iter.Iterator[string] {
+	combined := fmt.Sprintf("%s=%s", kv.Name, kv.Value)
+	t.buf = append(t.buf[:0],
+		kv.Name,
+		t.prefix+kv.Name,
```

Comment:
The token is used to check membership in the series, the prefix+token is used to identify membership in a specific chunk.

Since the bloom is for the full series, but at query time we need to filter out chunks, we do the following:

1. Test membership of `token` in the bloom filter
2. If negative, then `token` is not present in any chunk of the series -> filter out all chunks from series
3. If positive, then iterate over all input chunks, prefix the `token` with the chunk prefix (from/through/checksum) and test membership of `prefix+token` in the bloom filter -> filter out individual chunks from the series

Reply:
We can discuss whether we need
* `kv.Name`
* `t.prefix+kv.Name`
* `kv.Value`
* and `t.prefix+kv.Value`

or, if we could simply index
* `kv.Name+"="+kv.Value`
* and `t.prefix+kv.Name+"="+kv.Value`

Reply:
Right